README.md
# XCodeGen
a simple pattern code generator written in Haskell

## example1
```haskell
{-# LANGUAGE OverloadedStrings #-}

import Text.Gen
import Text.Gen.C

main :: IO ()
main = do
    putStrLn $ unpack $ runGen "    " $ do
        include' "stdio.h"
        define "PI" "3.1415926" 
        newline
        enum "Index" $ raw `fmap` ["first","second","third","forth","fifth"] 
        newline 
        struct "Fuck" $ do
           statement $ var int "a"
           statement $ var char "c"
        newline
        function char "nmsl" [var char "asd",var int "abc"]
        cblock $ do
            statement $ do
                raw "asd"
                binaryOp "=" $ do
                    raw "asd"
                    binaryOp "+" $ raw "abc"
            statement $ raw "return asd"
        newline
        function int "main" [var int "argc",var char "argv[]"]
        cblock $ do
            statement $ call "nmsl" [char' 'a',number 123]
            statement $ call "printf" [call "nmsl" [char' 'b',number 222]]
            statement $ raw "return 0"
```
output:
```c++
#include "stdio.h"
#define PI 3.1415926

enum Index{first,second,third,forth,fifth};

struct Fuck{
    int a;
    char c;
};

char nmsl(char asd,int abc){
    asd = asd + abc;
    return asd;
}

int main(int argc,char argv[]){
    nmsl('a',123);
    printf(nmsl('b',222));
    return 0;
}
```

## example2 
```haskell
{-# LANGUAGE OverloadedStrings #-}

import Text.Gen
import Text.Gen.Json

main :: IO ()
main = do
    putStrLn $ unpack $ runGen "    " $ do
        object $ do
            lineComma $ value "name" $ string "john"
            lineComma $ value "age" $ number 3
            lineComma $ value "ids" $ array $ string `fmap` ["asd","ads"]
            line $ value "info" $ object $ do
                lineComma $ value "asd" $ string "asd"
                lineComma $ value "bl" $ bool False
                line $ value "asdsa" $ string "asd"
```
output:
```json
{
    "name":"john",
    "age":3,
    "ids":["asd","ads"],
    "info":{
        "asd":"asd",
        "bl":false,
        "asdsa":"asd"
    }
}
```

ChangeLog.md
# Changelog for tmp

## Unreleased changes


src/Text/Gen.hs
{-# LANGUAGE OverloadedStrings #-}

module Text.Gen(
    Gen,
    runGen,
    raw,
    line,
    newline,
    indent,
    pair,
    split,
    block
)where

import Control.Monad
import Control.Monad.Writer.Lazy
import Control.Monad.State.Lazy
import Control.Monad.Reader
import Data.Monoid
import Data.Text hiding (split)

--State is indent
--Reader is indent character(Text,u can use 4 spaces to replace tab)
--Writer is Raw string
type Gen a = StateT Int (ReaderT Text (Writer Text)) a

for :: (Integral i,Monad m) => i -> (i -> m a) -> m ()
for i f = 
    if i /= 0 then do
        f i
        for (i-1) f
    else
        return ()


runGen :: Text -> Gen () -> Text
runGen indc g =
    let r = evalStateT g 0
        w = runReaderT r indc
    in execWriter w

raw :: Text -> Gen ()
raw = tell

newline :: Gen ()
newline = tell "\n"

line :: Gen () -> Gen ()
line act = do
    indent
    act
    newline

indent :: Gen ()
indent = do
    ind <- get
    indc <- ask
    for ind $ \_ -> tell indc

pair :: Text -> Text -> Gen () -> Gen ()
pair left right act = do
    tell left
    act
    tell right

split :: Text -> [Gen ()] -> Gen ()
split _ [] = return ()
split _ (x:[]) = x
split t (x:xs) = do
    x
    tell t
    split t xs

block :: Text -> Text -> Gen () -> Gen ()
block front back act = do
    raw front
    ind <- get
    put $ ind + 1
    act
    put $ ind
    indent
    raw back



src/Text/Gen/C.hs
{-# LANGUAGE OverloadedStrings #-}

module Text.Gen.C(
    int,
    uint,
    short,
    ushort,
    long,
    ulong,
    longlong,
    ulonglong,
    unsigned,
    char,
    uchar,
    float,
    double,
    void,

    statement,
    var,
    function,
    cblock,
    number,
    char',
    string,
    call,
    return',
    binaryOp,
    if',
    else',
    while,
    for,
    include,
    include',
    define,
    struct,
    union,
    enum
)where

import Data.Text hiding (split)
import Text.Gen 

--some userful constant
int :: Text 
int = "int"

uint :: Text 
uint = "unsigned int"

short :: Text
short = "short"

ushort :: Text
ushort = "unsigned short"

long :: Text
long = "long"

ulong :: Text
ulong = "unsigned long"

longlong :: Text
longlong = "long long"

ulonglong :: Text
ulonglong = "unsigned long long"

unsigned :: Text
unsigned = "unsigned" 

char :: Text
char = "char"

uchar :: Text
uchar = "unsigned char"

float :: Text
float = "float"

double :: Text
double = "double"

void :: Text
void = "void"

--functions 

statement :: Gen () -> Gen ()
statement act = do
    indent
    act
    raw ";\n"

var :: Text -> Text -> Gen ()
var typ name = do
    raw $ typ `append` (' ' `cons` name)

function :: Text -> Text -> [Gen ()] -> Gen ()
function rettype name args = do
    var rettype name
    pair "(" ")" $ split "," args

cblock :: Gen () -> Gen ()
cblock acts = block "{\n" "}\n" acts

number :: (Show n,Num n) => n -> Gen ()
number num = raw $ pack $ show num

char' :: Char -> Gen ()
char' ch = pair "\'" "\'" $ raw $ singleton ch

string :: Text -> Gen ()
string str = pair "\"" "\"" $ raw str

call :: Text -> [Gen ()] -> Gen ()
call name args = pair (name `append` "(") ")" $ split "," args

return' :: Gen () -> Gen ()
return' act = statement $ do
    raw "return "
    act

binaryOp :: Text -> Gen () -> Gen ()
binaryOp op act = do
    raw " "
    raw op
    raw " "
    act

if' :: Gen () -> Gen ()
if' pred = pair "if(" ")" pred

else' :: Gen ()
else' = raw "else"

for :: [Gen ()] -> Gen ()
for (a:b:c:xs) = do
    pair "for(" ")" $ do
        a
        raw ";"
        b
        raw ";"
        c

while :: Gen () -> Gen ()
while pred = pair "while(" ")" pred

--dowhile

include :: Text -> Gen ()
include file = do
    raw "#include "
    pair "<" ">" $ raw file
    newline 

include' :: Text -> Gen ()
include' file = do
    raw "#include "
    string file
    newline

define :: Text -> Text -> Gen ()
define name context = do
    raw "#define "
    raw name
    raw " "
    raw context
    newline

--

struct :: Text -> Gen () -> Gen ()
struct name act = do
    raw "struct "
    raw name
    block "{\n" "};" act
    newline

union :: Text -> Gen () -> Gen ()
union = struct

enum :: Text -> [Gen ()] -> Gen ()
enum name acts = do
    raw "enum "
    raw name
    pair "{" "};" $ split "," acts
    newline



app/Main.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Text.Gen
--import Text.Gen.Json
--import Text.Gen.C
import Text.Gen.Latex
import Data.Text
import System.IO

--main :: IO ()
--main = do
--    putStrLn $ unpack $ runGen "    " $ do
--        block "{\n" "}\n" $ do
--            line $ raw "asd;"
--            line $ do
--                raw "func"
--                pair "(" ");" $ Text.Gen.split "," [raw "a",raw "b"]
--            line $ raw "asd"
--            line $ do
--                raw "if"
--                pair "(" ")" $ raw "a == b"
--                block "{\n" "}\n" $ do
--                    line $ raw "asd;"
--                    line $ raw "adsasd;"
--
-- main :: IO ()
-- main = do
--     putStrLn $ unpack $ runGen "    " $ do
--         object $ do
--             lineComma $ value "name" $ string "john"
--             lineComma $ value "age" $ number 3
--             lineComma $ value "ids" $ array $ string `fmap` ["asd","ads"]
--             line $ value "info" $ object $ do
--                 lineComma $ value "asd" $ string "asd"
--                 lineComma $ value "bl" $ bool False
--                 line $ value "asdsa" $ string "asd"

--main :: IO ()
--main = do
--    putStrLn $ unpack $ runGen "    " $ do
--        include' "stdio.h"
--        define "PI" "3.1415926" 
--        newline
--        enum "Index" $ raw `fmap` ["first","second","third","forth","fifth"] 
--        newline 
--        struct "Fuck" $ do
--           statement $ var int "a"
--           statement $ var char "c"
--        newline
--        function char "nmsl" [var char "asd",var int "abc"]
--        cblock $ do
--            statement $ do
--                raw "asd"
--                binaryOp "=" $ do
--                    raw "asd"
--                    binaryOp "+" $ raw "abc"
--            statement $ raw "return asd"
--        newline
--        function int "main" [var int "argc",var char "argv[]"]
--        cblock $ do
--            statement $ call "nmsl" [char' 'a',number 123]
--            statement $ call "printf" [call "nmsl" [char' 'b',number 222]]
--            statement $ raw "return 0"

main :: IO ()
main = do
    putStrLn $ unpack $ runGen "    " $ do
        aligned $ do
            indent
            sqrt_ $ (raw "3") `frac` (raw "4")
            newline_
